google/resource_kms_crypto_key_test.go

```diff
@@ -0,0 +1,381 @@
+		}
+
+		keyRingId, err := parseKmsKeyRingId(rs.Primary.Attributes["key_ring"], config)
+
+		if err != nil {
+			return err
+		}
+
+		cryptoKeyId := &kmsCryptoKeyId{
+			KeyRingId: *keyRingId,
+			Name:      rs.Primary.Attributes["name"],
+		}
+
+		listCryptoKeysResponse, err := config.clientKms.Projects.Locations.KeyRings.CryptoKeys.List(cryptoKeyId.parentId()).Do()
+		if err != nil {
+			return fmt.Errorf("Error listing KeyRings: %s", err)
+		}
+
+		for _, cryptoKey := range listCryptoKeysResponse.CryptoKeys {
+			log.Printf("[DEBUG] Found CryptoKey: %s", cryptoKey.Name)
+
+			if cryptoKey.Name == cryptoKeyId.cryptoKeyId() {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("CryptoKey not found: %s", cryptoKeyId.cryptoKeyId())
+	}
+}
+
+func testAccCheckGoogleKmsCryptoKeyHasRotationParams(rotationPeriod, resourceName string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		config := testAccProvider.Meta().(*Config)
+
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("Resource not found: %s", resourceName)
+		}
+
+		keyRingId, err := parseKmsKeyRingId(rs.Primary.Attributes["key_ring"], config)
+
+		if err != nil {
+			return err
+		}
+
+		cryptoKeyId := &kmsCryptoKeyId{
+			KeyRingId: *keyRingId,
+			Name:      rs.Primary.Attributes["name"],
+		}
+
+		getCryptoKeyResponse, err := config.clientKms.Projects.Locations.KeyRings.CryptoKeys.Get(cryptoKeyId.cryptoKeyId()).Do()
+
+		if err != nil {
+			return err
+		}
+
+		if rotationPeriod != getCryptoKeyResponse.RotationPeriod {
+			return fmt.Errorf("Expected rotation period %s to match input %s", getCryptoKeyResponse.RotationPeriod, rotationPeriod)
+		}
+
+		_, err = time.Parse(time.RFC3339Nano, getCryptoKeyResponse.NextRotationTime)
+
+		return err
+	}
+}
+
+/*
+	KMS KeyRings cannot be deleted. This ensures that the CryptoKey resource was removed from state,
+	even though the server-side resource was not removed.
+*/
+func testAccCheckGoogleKmsCryptoKeyWasRemovedFromState(resourceName string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		_, ok := s.RootModule().Resources[resourceName]
+
+		if ok {
+			return fmt.Errorf("Resource was not removed from state: %s", resourceName)
+		}
+
+		return nil
+	}
+}
+
+/*
+	KMS KeyRings cannot be deleted. This ensures that the CryptoKey resource's CryptoKeyVersion
+	sub-resources were scheduled to be destroyed, rendering the key itself inoperable.
+*/
+
+func testAccCheckGoogleKmsCryptoKeyVersionsDestroyed(projectId, location, keyRingName, cryptoKeyName string) resource.TestCheckFunc {
+	return func(_ *terraform.State) error {
+		config := testAccProvider.Meta().(*Config)
+		gcpResourceUri := fmt.Sprintf("projects/%s/locations/%s/keyRings/%s/cryptoKeys/%s", projectId, location, keyRingName, cryptoKeyName)
+
+		response, _ := config.clientKms.Projects.Locations.KeyRings.CryptoKeys.CryptoKeyVersions.List(gcpResourceUri).Do()
```

Comment:
I'd feel more comfortable if this error were handled: if we got a mysterious server error, then `response.CryptoKeyVersions` would (I believe) be empty, which means this test would succeed, even though we don't _actually_ know that the versions were destroyed.